dockstore-client/src/main/java/io/dockstore/client/cli/nested/SecondaryFilesUtility.java

```diff
@@ -0,0 +1,208 @@
+        return inputFileIds;
+    }
+
+    private void loopThroughSource(String id, String fileId, LinkedTreeMap mapStep, List<Map<String, List<String>>> idAndSecondaryFiles) {
+        String[] temp = id.split("#");
+        String temp3 = temp[temp.length - 1];
+        String[] split = temp3.split("/");
+        String toolIDFromWorkflow = split[split.length - 1];
+
+        String descriptor = mapStep.get("run").toString();
+        final String toolDescriptor = this.cwlUtil.parseCWL(descriptor).getLeft();
+        Object toolDescriptorObject;
+        try {
+            toolDescriptorObject = this.gson.fromJson(toolDescriptor, CommandLineTool.class);
+            if (toolDescriptorObject != null) {
+                CommandLineTool commandLineTool = (CommandLineTool)toolDescriptorObject;
+                List<CommandInputParameter> inputs = commandLineTool.getInputs();
+                inputs.forEach(input -> {
+
+                    try {
+                        @SuppressWarnings("unchecked")
+                        List<String> secondaryFiles = (List<String>)input.get("secondaryFiles");
+                        if (secondaryFiles != null) {
+                            String toolId = input.getId().toString();
+                            String[] temp2 = toolId.split("#");
+                            String toolFileId = temp2[1];
+                            // Check if the tool descriptor has secondary files and if the id matches the workflow id
+                            if (toolFileId.equals(toolIDFromWorkflow)) {
+                                //                            if (toolFileId.equals(workflowFileId)) {
+                                Map<String, List<String>> hashMap = new HashMap<>();
+                                hashMap.put(fileId, secondaryFiles);
+                                idAndSecondaryFiles.add(hashMap);
+                            }
+                        }
+                    } catch (ClassCastException e) {
+                        LOG.error(e.getMessage());
+                    }
+                });
+
+            }
+        } catch (JsonParseException ex) {
+            LOG.error("The JSON file provided is invalid.");
+        }
+    }
+
+    /**
+     * The modifies idAndSecondaryFiles to include secondaryFiles present in the tool descriptor (even though it may exist in the workflow descriptor)
+     *
+     * @param workflow            The workflow descriptor
+     * @param fileId              The current file ID we're looking at in the workflow descriptor
+     * @param idAndSecondaryFiles A file ids and secondary files that needed to be added to those file ids
+     */
+    private void getDescriptorsWithFileInput(Workflow workflow, String fileId, List<Map<String, List<String>>> idAndSecondaryFiles) {
+        Object steps = workflow.getSteps();
+        if (steps instanceof ArrayList) {
```

Comment:
In these loops, don't rely on the concrete class (ArrayList instead of List). 
You don't actually care and it makes the code longer